### DISCUSSION TITLE
`sortable()` by related record

### DISCUSSION THREAD

**Discussion by titantwentyone** [filamentphp/filament#1354] [Help]
I just wondered if anyone had any guidance (or indeed, if it is even possible) to sort a table via a related field. The following results in a SQL error:
```
Columns\TextColumn::make('customer.name')
    ->sortable();
```
where "customer" is a related record (in this case a morph).

Is there any way to override the query used when sorting by a particular column?

**Comment by danharrin** [comment]
This should work, does it work for more simple relationships like BelongsTo?

**Reply by titantwentyone** [reply]
Yeh. In another part of my app I have a parent/child  `belongsTo` relationship (literally a parent and their child):

```
Columns\TextColumn::make('parent.name')
    ->sortable()
```
And this works fine. I missed this before so assumed it wasn't working for any related records. It feels like it is the nature of the relationship being a morph that is problematic possibly?

If it helps, the specific error is:

```
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'name' in 'field list' (SQL: select * from `orders` where `orders`.`deleted_at` is null order by (select `name` from `orders` as `laravel_reserved_0` where `laravel_reserved_0`.`` = `orders`.`customer_id` and `laravel_reserved_0`.`deleted_at` is null) asc limit 10 offset 0) 
```

The table is being populated with the correct data - it's just when attempting to sort.

**Reply by danharrin** [reply]
Can you work out how to make it work in normal Eloquent? Then we can push the appropriate fix for Filament

**Reply by titantwentyone** [reply]
Thanks Dan. The follwing provides order details plus morph relationship data:

```
Order::query()
    ->with(['customer' => function(MorphTo $morphTo) {
            $morphTo->morphWith([
                ParentUser::class => [],
                SchoolUser::class => []
            ]);
        }
    ])->get();
```

**Reply by danharrin** [reply]
Sorry, where is the order clause applied here?

**Reply by titantwentyone** [reply]
Have updated:

```
Order::query()
    ->with(['customer' => function(MorphTo $morphTo) {
            $morphTo->morphWith([
                ParentUser::class => [],
                SchoolUser::class => []
            ]);
        }
    ])
    ->get()
    ->sortBy('customer.name'); // or sortByDesc('customer.name')
```

**Reply by danharrin** [reply]
This isnt a database sort though - you'd need to use `orderBy()` somehow, otherwise the sorting is done by PHP which is slow.

**Reply by ryangjchandler** [reply]
@titantwentyone You could potentially use a sub-select and order on that field.

**Reply by titantwentyone** [reply]
> This isnt a database sort though - you'd need to use `orderBy()` somehow, otherwise the sorting is done by PHP which is slow.

Yeh. I hunted around for a solution to that but all roads led to sorting the end result. I'll knock up a sub-select like @ryangjchandler suggested.

**Reply by titantwentyone** [reply]
Hoping this is it. Solution was to union the two "customer" tables and use the result in sub-select.

```
$customers = DB::table('school_users')
    ->select('name', 'id')
    ->addSelect(DB::raw('"App\\\Models\\\SchoolUser" as type'))
    ->union(DB::table('parent_users')
        ->select('name', 'id')
        ->addSelect(DB::raw('"App\\\Models\\\ParentUser" as type'))
    );

$result = DB::table('orders')
    ->joinSub($customers, 'customers', function($join) {
        $join->on('orders.customer_id', '=', 'customers.id');
        $join->on('orders.customer_type', '=', 'customers.type');
    })
    ->select('customers.id as customer_id', 'customers.name', 'customers.type', 'orders.customer_type', 'orders.id', 'amount')
    ->orderBy('customers.name');
```

**Reply by danharrin** [reply]
Not sure how we can transform this to be a solution that we can apply to all columns though, its super ugly :(

**Reply by titantwentyone** [reply]
It's massively ugly. I totally agree. I couldn't find a more succinct way of doing this - that's not to say there isn't one though! I did look at `whereHasMorph` but I can only get `Order` data from that and not pull in the related data AFAIK.

If you've got any ideas of avenues I can explore from here, I'd be more than happy to investigate. I'm still going to look into this peridoically but any flashes of inspiration would be most welcome!

**Reply by titantwentyone** [reply]
I've really struggled trying to find a more elegant solution than the one provided. I think at this stage, I'm going to have to make "special case" of this particular column. Is there a sane way to apply the sort to the table? My inital thoughts are to extend `Column` and override `applySort` from the `InteractsWithTableQuery` trait returning the appropriate query based on the column name. Is this an "okay" approach. Any better solution? This clearly would a special case at this stage without any way of applying it in a more uniform way.

**Reply by danharrin** [reply]
Yeah I agree with that, or let the column handle its own sorting and define a `sortUsing()` method or something, which accepts a closure.

**Reply by francoism90** [reply]
@danharrin We have the same problem, but couldn't find the answer yet:
```php
TextColumn::make('model.name')
            ->label(__('Organisation'))
            ->sortable();
```
The model is a `morphTo` relationship.

Results in: `Column not found: 1054 Unknown column 'laravel_reserved_0.' in 'where clause'`

Is this expected? Thanks!

**Reply by danharrin** [reply]
How would you order by a morphto relationship in normal Eloquent?

**Reply by francoism90** [reply]
@danharrin Can I use this method: https://github.com/kirschbaum-development/eloquent-power-joins?tab=readme-ov-file#3---order-by

It seems Filament is using this to build the relation.

**Reply by danharrin** [reply]
I do not think that covers MorphTo, where many different tables need to be queried to build the order.

Filament is not using that package to render table content. Only summaries and some form components/filters.

**Reply by francoism90** [reply]
@danharrin Ok, thanks for looking into this! :)
